Feature/minyeong

NyeongDev opened the pull request:
## 요구 사항

https://user-images.githubusercontent.com/110284486/225786921-c85b515c-c6c5-4af9-be7f-3b5d21182e32.mp4

- [X] mock data의 key값(시간)을 기반으로 시계열 차트 구현
- [X] Area, Bar 형태가 합쳐진 복합 그래프 형태
- [X] 데이터 구역에 마우스 호버 시 id, value_area, value_bar 데이터를 툴팁 형태로 제공
- [X] 버튼과 데이터 클릭 시 특정 데이터를 하이라이트 하는 필터링 기능 구현

<br/>

## 구현 방법
- 커스텀이 자유로운 apexcharts 라이브러리 사용
- API 호출 및 응답 데이터 가공 처리 함수는 useChartData() custom hook으로 구현
- chartTypes.ts 파일 분리
- 차트 페이지와 차트, 툴팁 컴포넌트를 분리하여 UI 수정이 용이하도록 함
- fillColor 속성을 사용하여 선택된 데이터의 색 변경

<br/>

## 궁금한 사항
1. ChartPage에서 useChartData 훅을 사용하여 데이터를 가져오고, 필터 기능 때문에 한번 더 useState에 데이터를 저장하고 있습니다. 즉 훅에서도 한 번, 페이지에서도 한 번 useState를 사용하고 있는데 혹시 불필요하게 useState를 사용하고 있는 부분이 있는지 궁금합니다.

2. area 차트 영역을 하이라이트 하는 방법을 찾지 못해 마커 색을 변경하는 방식을 적용하게 되었습니다. 혹시 알고 계신 방법이 있다면 공유 부탁드려요!

3.  타입을 다른 파일로 분리할 때 어떤 기준을 적용하는 것이 좋을지 궁금합니다. 같은 파일 안에 있을 때 가독성이 더 좋아지는 경우도 있는 것 같아서 궁금했습니다! 여러분은 어떻게 하고 계신가요?

<br/>

## 추후 보완 사항

- [x] 필터 기능에 라우팅을 적용하여 페이지 이동 시, url로 차트 공유 시 동일하게 필터링 된 차트를 확인할 수 있도록 하기
- [x] css 작업
  - 특정 데이터가 하이라이트 되었을 때 해당하지 않는 데이터 색상을 흑백으로 변경
  - 버튼 컴포넌트 분리 및 url에 따라 버튼 색상이 변경되도록 하여 현재 어떤 지역의 데이터를 하이라이트 하고 있는지 확인하기 쉽도록 함
  - 헤더 추가
- [ ] 필터링 로직을 hook으로 분리하여 유지보수성 향상

NyeongDev committed:
[#2] Feat : 차트 구현

- mock data의 key값(시간)을 기반으로 시계열 차트 구현
- Area, Bar 형태가 합쳐진 복합 그래프 형태
- 데이터 구역에 마우스 호버 시 id, value_area, value_bar 데이터를 툴팁 형태로 제공
- api 모듈 분리
- chartTypes.ts 파일 분리

NyeongDev committed:
[#2] Refactor : 툴팁 컴포넌트 분리, useChartData 훅 구현

- 툴팁 컴포넌트를 분리하여 UI 수정이 용이하도록 함
- useChartData 훅 구현, response data를 Chart에서 사용할 수 있는 형태로 가공
- 사용하지 않는 코드 삭제

NyeongDev committed:
[#2] Feat : 필터링 기능 구현

- 버튼과 데이터 클릭 시 특정 데이터를 하이라이트 하는 방식
- fillColor 속성을 사용하여 선택된 데이터의 색 변경
- new Set 문법을 사용하기 위해 tsconfig.json es6로 수정
- ChartPage에서 필터 기능 구현, 필터링 된 데이터를 Chart 컴포넌트에 props로 전달하도록 변경

whl5105 commented:
**궁금한 사항 답변**
1. 저는 선택된 차트 필터링 값을 바로 query string으로 반영해서 별도로 state 선언이 필요 없었습니다! 3곳에서 동일한 데이터가 state 되는 건 말씀하신 대로 조금 불필요한 선언이 아닐까 생각합니다! 코드 리뷰 때 자세하게 이야기해봐요!
2. 하이라이트는 저도 pointer 컬러를 변경해 주었어요! segment로 분리를 해보니 선택된 값의 전후 값의 범위가 전체로 잡히더라고요. 다른 방법 알게되시면 함께 공유해요 !! 
3. 타입의 경우 Response 타입만 type 파일로 선언해 주었고 각 컴포넌트 타입들은 파일 안에서 확인할 수 있도록 선언해 주었습니다! props 타입들은 다른 컴포넌트에서 사용하는 타입이 아니기 때문에 같은 파일 안에 있을 때 가독성이 더 좋아 보입니다!

민영님도 apexcharts로 구현하셨다니.. 대단하십니다! 고생하셨어요 ! 👍

minhyeonhong commented:
2. legend속성과 series의 name같으면 저절로 하이라이트 기능이 되던데.. 이번 과제는 legend가 series보다 많아서 해당 기능을 못 쓰겠더라구요 ㅠㅠ

series data속성에 color를 줄수 있었군요 배워갑니다...

NyeongDev commented:
> **궁금한 사항 답변**
> 
> 1. 저는 선택된 차트 필터링 값을 바로 query string으로 반영해서 별도로 state 선언이 필요 없었습니다! 3곳에서 동일한 데이터가 state 되는 건 말씀하신 대로 조금 불필요한 선언이 아닐까 생각합니다! 코드 리뷰 때 자세하게 이야기해봐요!
> 2. 하이라이트는 저도 pointer 컬러를 변경해 주었어요! segment로 분리를 해보니 선택된 값의 전후 값의 범위가 전체로 잡히더라고요. 다른 방법 알게되시면 함께 공유해요 !!
> 3. 타입의 경우 Response 타입만 type 파일로 선언해 주었고 각 컴포넌트 타입들은 파일 안에서 확인할 수 있도록 선언해 주었습니다! props 타입들은 다른 컴포넌트에서 사용하는 타입이 아니기 때문에 같은 파일 안에 있을 때 가독성이 더 좋아 보입니다!
> 
> 민영님도 apexcharts로 구현하셨다니.. 대단하십니다! 고생하셨어요 ! 👍

@whl5105
재사용하는 타입인 경우 type 파일에 선언하고, 각 컴포넌트에서만 사용하거나 props 타입일 경우 해당 파일 안에서 선언한다고 이해했습니다! 좋은 방법인 것 같아요 감사합니다!😀

eunrain commented:
와 apexchart로 완성하셨다니.. 너무 대단하십니다 👍
area 그래프 마커 색 하이라이트로 대신하신 점 인상 깊었습니다! 생각도 못했어요!
저는 사용하고 있는 타입을 모두 type 폴더에서 한꺼번에 관리해주고 있습니다. 한꺼번에 관리해줌으로써 코드가 깔끔해 보이긴 하지만, 가독성은 떨어지는 것도 같습니다. 이 부분도 코드리뷰때 같이 얘기 나눠보면 좋을 것 같아요! 
고생하셨습니다!!!

yeon-kk reviewed (commented):
CharPage의 handleFilter 함수에서, map메서드 코드가 중복되어서 줄여볼 수도 있을 것 같아요.
말씀해주신 state 부분도 이야기해보면 좋을것 같아요.
툴팁을 따로 컴포넌트로 분리하셔서 코드를 읽기 편해요
타입 폴더에 해당 컴포넌트와 관련된 타입을 관련된 파일에 모아두셨는데, 찾아보기 편해서 좋은 방법이라고 생각합니다. 다시 찾을 때 찾기 편한 방법이 있으시다면 그 방법으로 하시는것도 좋다고 생각해요

Hwang-Junsu commented:
**궁금한 사항 답변**
1. 저는 필터링을 chart 내부에서 함수를 짜서 state가 필요하지 않았습니다. hooks 내에서 데이터를 변경하고 싶다면 state가 필요할 수 도 있겠다는 생각이 드네요. 여러가지 방법들을 시도하고 고민하는것이 좋을 것 같습니다!

2. 저도 area차트를 하이라이트하는 방법을 찾지 못해서 고민을 했습니다...😂 그래서 bar차트를 반투명하게 변경하면서 area chart도 변경되는 것처럼 보일 수 있도록 하였습니다.

3. 재사용 가능한 타입의 경우에는 분리하고, 아니라면 컴포넌트 상단에 선언하는 것도 괜찮다고 생각합니다. 방법이 정해진 건 없는 것 같아요!

고생하셨습니다!!

NyeongDev committed:
[#2] Feat : 필터 기능에 라우팅 적용

- useParams 이용

NyeongDev committed:
[#2] Style : 차트 css 작업

- 특정 데이터가 하이라이트 되었을 때 해당하지 않는 데이터 색상을 흑백으로 변경
- 버튼 컴포넌트 분리 및 url에 따라 버튼 색상이 변경되도록 하여 현재 어떤 지역의 데이터를 하이라이트 하고 있는지 확인하기 쉽도록 함
- 헤더 추가

NyeongDev committed:
[#2] Style : 툴팁 css 작업

eunrain commented:
민영님은 useparams를 사용해서 필터링 구현하셨군요..! 코드가 훨씬 깔끔해진 것 같아 보기 좋습니다! 고생하셨어요 👍
